### PR TITLE
.Net: Introduce IAIServiceSelector interface to allow AI service and request settings to be selected during semantic function execution

### PR DESCRIPTION
**Pull request by markwallace-microsoft**
### Motivation and Context

Fix following issues for Semantic functions:

1. Allow semantic functions to dynamically retrieve the AI service and request settings to be used when executing the function

Previously semantic functions execution has the following limitations:

1. The AI service was set on the semantic function when it was created. This meant that an instance of a semantic function could not be used with multiple different Kernel instances.
1. The AI request settings were set on the semantic function when it was created. This meant that different request settings could not be selected when the function was executed.

### Description

This PR add's a new abstraction called `IAIServiceSelector`. An instance of this is added to a semantic function when it is created. The default implementation works as follows:

1. If the semantic function only has a single associated model request setting instance with no service id then the default AI service and the model request settings are used. This is consistent with the previous behaviour.
1. If the semantic function only has a single associated model request setting instance with a service id then the named AI service and the model request settings are used. If the named AI service is not available an exception will be thrown.  This is consistent with the previous behaviour.
1. If the semantic function only has multiple associated model request setting instances then:
    1. The model request setting instances are considered in order
    1. The first model request setting instance with no service id is considered the default
    1. For each model request setting instance that has a service id, if the service exists then it will be used with the associated model request settings
    1. If no matching service can be found and default request settings are provided the default service will be used
    1. If no matching service can be found and no default request settings are provided an exception will be thrown

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/SemanticKernel.Abstractions/Functions/IAIServiceSelector.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+{
+    /// <summary>
+    /// Return the AI service and requesting settings from the specified provider based on the model settings.
+    /// The returned value is a tuple containing instances of <see cref="IAIService"/> and <see cref="AIRequestSettings"/>
+    /// </summary>
+    /// <typeparam name="T">Type of AI service to return</typeparam>
+    /// <param name="serviceProvider">AI service provider</param>
+    /// <param name="modelSettings">Collection of model settings</param>
+    /// <returns></returns>
+    (T?, AIRequestSettings?) SelectAIService<T>(IAIServiceProvider serviceProvider, List<AIRequestSettings>? modelSettings) where T : IAIService;
```
Comment (SergeyMenshykh) [inline dotnet/src/SemanticKernel.Abstractions/Functions/IAIServiceSelector.cs:24]: nit: IReadOnlyList?

File: dotnet/src/SemanticKernel.Core/Functions/OrderedIAIServiceSelector.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            {
+                return (service, modelSettings?.FirstOrDefault<AIRequestSettings>());
+            }
+        }
+        else
+        {
+            AIRequestSettings? defaultRequestSettings = null;
+            foreach (var model in modelSettings)
+            {
+                if (model.ServiceId is not null)
```
Comment (SergeyMenshykh) [inline dotnet/src/SemanticKernel.Core/Functions/OrderedIAIServiceSelector.cs:34]: Consider checking for not empty string as well.

File: dotnet/src/SemanticKernel.Core/Functions/OrderedIAIServiceSelector.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            }
+
+            if (defaultRequestSettings is not null)
+            {
+                return (serviceProvider.GetService<T>(null), defaultRequestSettings);
+            }
+        }
+
+        var names = string.Join("|", modelSettings.Select(model => model.ServiceId).ToArray());
+        throw new SKException($"Service of type {typeof(ITextCompletion)} and name {names ?? "<NONE>"} not registered.");
```
Comment (SergeyMenshykh) [inline dotnet/src/SemanticKernel.Core/Functions/OrderedIAIServiceSelector.cs:56]: I believe you wanted to include type of the T into the exception message and not of ITextCompletion?

File: dotnet/src/SemanticKernel.Abstractions/Functions/ISKFunction.cs [hunk 4]
```diff
@@ -34,9 +35,9 @@ public interface ISKFunction
     string Description { get; }
 
     /// <summary>
-    /// AI service settings
+    /// Model request settings.
     /// </summary>
-    AIRequestSettings? RequestSettings { get; }
+    List<AIRequestSettings>? ModelSettings { get; }
```
[the remark this answers is not in the document]
Comment (markwallace-microsoft) [reply dotnet/src/SemanticKernel.Abstractions/Functions/ISKFunction.cs]: Yep, I have obsoleted the `RequestSettings` property and am not replacing it.

**Review by SergeyMenshykh — APPROVED** [review]

File: dotnet/src/SemanticKernel.Core/Functions/OrderedIAIServiceSelector.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+/// <summary>
+/// Implementation of <see cref="IAIServiceSelector"/> that selects the AI service based on the order of the model settings.
+/// Uses the service id to select the preferred service provider and then returns the service and associated model settings.
+/// </summary>
+internal class OrderedIAIServiceSelector : IAIServiceSelector
+{
+    /// <inheritdoc/>
+    public (T?, AIRequestSettings?) SelectAIService<T>(IAIServiceProvider serviceProvider, IReadOnlyList<AIRequestSettings>? modelSettings) where T : IAIService
+    {
+        if (modelSettings == null || modelSettings.Count == 0)
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.Core/Functions/OrderedIAIServiceSelector.cs:20]: nit
```suggestion
        if (modelSettings is null || modelSettings.Count == 0)
```

File: dotnet/src/SemanticKernel.Core/Functions/OrderedIAIServiceSelector.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        {
+            var service = serviceProvider.GetService<T>(null);
+            if (service != null)
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.Core/Functions/OrderedIAIServiceSelector.cs:23]: nit
```suggestion
            if (service is not null)
```

File: dotnet/src/SemanticKernel.Core/Functions/OrderedIAIServiceSelector.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        }
+        else
+        {
+            AIRequestSettings? defaultRequestSettings = null;
+            foreach (var model in modelSettings)
+            {
+                if (!string.IsNullOrEmpty(model.ServiceId))
+                {
+                    var service = serviceProvider.GetService<T>(model.ServiceId);
+                    if (service != null)
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.Core/Functions/OrderedIAIServiceSelector.cs:36]:
```suggestion
                    if (service is not null)
```

File: dotnet/src/SemanticKernel.Core/Functions/OrderedIAIServiceSelector.cs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+                else
+                {
+                    // First request settings with empty or null service id is the default
+                    defaultRequestSettings ??= model;
+                }
+            }
+
+            if (defaultRequestSettings is not null)
+            {
+                return (serviceProvider.GetService<T>(null), defaultRequestSettings);
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.Core/Functions/OrderedIAIServiceSelector.cs:50]: Do we allow the `SelectAIService` to return null from this `GetService<T>` call?

The other scenarios we throw, thinking about consistency suggest throwing in here as well.

File: dotnet/src/SemanticKernel.Core/Functions/SemanticFunction.cs [hunk 9]
```diff
@@ -72,9 +72,12 @@ internal sealed class SemanticFunction : ISKFunction, IDisposable
             description: promptTemplateConfig.Description,
             pluginName: pluginName,
             functionName: functionName,
+            serviceSelector: serviceSelector,
             loggerFactory: loggerFactory
-        );
-        func.SetAIConfiguration(promptTemplateConfig.GetDefaultRequestSettings());
+        )
+        {
+            _modelSettings = promptTemplateConfig.ModelSettings
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.Core/Functions/SemanticFunction.cs:79]: Why does a PromptTemplateConfiguration has model settings ? This configuration should be passed directly to the `FromSemanticConfig()`.

PromptTemplateEngine, should not have any logic or knowledge of the `Model` settings.

**Review by RogerBarreto — APPROVED** [review]
Nit comments.
Suggestion on changing `SelectAIService` to aways throw if no service is found. + Removing any logic that handles ITextCompletion being null in the SemanticFunction InvokeAsync impl.

Overall LGTM

File: dotnet/src/SemanticKernel.Core/Functions/SemanticFunction.cs (hunk 9 above)
Comment (markwallace-microsoft) [reply dotnet/src/SemanticKernel.Core/Functions/SemanticFunction.cs:79]: `PromptTemplateConfig.cs` is the model object for a config.json. We allow the model settings to be set in the config.json.

**Review by RogerBarreto — APPROVED** [review]

**Pull request merged**